### PR TITLE
prosemirror-commands: Fix Command type

### PR DESCRIPTION
**Pull request by timjb**
The 'dispatch' and 'view' arguments must be optional to support
'dry runs'.

Also add doc comment for 'Command' type by copying relevant the part
from the ProseMirror documentation.

Also simplify type definition of 'chainCommands' using 'Command'.

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [ ] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: [ProseMirror docs](https://prosemirror.net/docs/ref/#commands), [discussion in PR which added the `Command` interface](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/45836#discussion_r452077254)
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [x] Include [tests for your changes](https://github.com/DefinitelyTyped/DefinitelyTyped#testing)
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

### CODE REVIEW

**Comment by mmorearty** [comment]
Following up on discussion from https://github.com/DefinitelyTyped/DefinitelyTyped/pull/45836

@aried3r is right that that's something we probably need to change; but I think there's a bit of subtlety to it. Commands are documented as having different behavior in different situations.

In general, a command, when called, should indeed expect that `dispatch` and/or `view` may not have been passed. So from the callee's perspective, the general signature would indeed be `(state, dispatch?, view?)`. This applies if type `Command` is to be used as a template for writing your own commands.

But it may be different from the caller's perspective — at least for some callers. prosemirror-keymap is explicitly [documented](https://prosemirror.net/docs/ref/#keymap) as always passing all three arguments to any command that you put into a keymap. "Bindings should map key names to command-style functions, which will be called with `(EditorState, dispatch, EditorView)` arguments."

Interestingly, that paragraph goes on to make a further point: "Note that the view argument isn't part of the command protocol, but can be used as an escape hatch if a binding needs to directly interact with the UI." Both prosemirror-keymap and prosemirror-commands mention the third `EditorView` argument, but apparently it's not considered part of the command protocol.

Importantly, the current `Command` type does allow you to take a function that only accepts one or two arguments, and put it into a keymap:

```ts
// This function only takes one arg, but you can assign it to a variable of type Command
const noop: Command = (state) => false

// dispatch arg is optional
const sayHello: Command = (state, dispatch?: (tr: Transaction) => void) => {
    const tr = state.tr.insertText("hello");
    if (dispatch) dispatch(tr);
    return true;
}

const plugin = keymap({
    "a": noop,
    "b": sayHello,
});
```

That's why the tests added in https://github.com/DefinitelyTyped/DefinitelyTyped/pull/45836 include this:

```tr
const keymap: commands.Keymap = {
    ArrowUp: () => true,  // takes no args
    ArrowDown: commands.deleteSelection,  // takes two args
    ArrowLeft: commands.joinBackward,  // takes three args
    ArrowRight: (state, dispatch, view) => true,  // arg types inferred
};
```

Since the docs say that prosemirror-keymap passes all three args to the functions in the keymap, I would think that if we want to be correct, then the function signatures in prosemirror-keymap/index.d.ts (plus `pcBaseKeymap`, `macBaseKeymap`, and `baseKeymap` in prosemirror-commands) should include all three args, because:

1. As shown above, because of covariance, you can still put commands that accept one or two arguments into it.
2. But it also enforces the documented requirement that anyone who is trying to call functions in a keymap must pass all three arguments. (But in practice, client code is very unlikely to do this.)

In conclusion, perhaps `Command` should be changed, so it can be used as a template for writing your own commands, but if so, perhaps prosemirror-keymap and the three keymaps in prosemirror-commands should not use that type, but should be more explicit. What do you think?

**Review by mmorearty — APPROVED** [review]
Despite my musings in my earlier comment, this seems better than what we had before. It allows people to use `Command` as a type for their own commands. And my point about the fact that keymaps are documented as always calling commands with all three arguments is probably moot, because ProseMirror is written in JavaScript, not TypeScript, and always will be. So, this looks good to me. Thanks @timjb!
